src/main/kotlin/com/cultureamp/eventsourcing/PGSessionAdvisoryBookmarkLock.kt

```diff
@@ -0,0 +1,77 @@
+package com.cultureamp.eventsourcing
+
+import org.slf4j.LoggerFactory
+import java.sql.Connection
+import java.util.UUID
+
+private val logger = LoggerFactory.getLogger(PGSessionAdvisoryBookmarkLock::class.java)
+class PGSessionAdvisoryBookmarkLock(
+    private val connectionProvider: () -> Connection,
+) : BookmarkLock {
+    private val lock = Any()
+    internal var connection = getConnection()
+    /**
+     * The set of bookmarks that we believe we hold. Note that this is only for diagnostic purposes,
+     * Postgres is the source of truth. Currently we just use this to detect when locks change hands
+     * so we can log it
+     */
+    private var locksHeld = mutableSetOf<String>()
```

Comment:
Hmm - this BookmarkLock is shared across all projector threads, right? I'm wondering if we need to worry about thread safety / reliability of this field, even if it's only used for logging. WDYT?

Reply:
Actually is there a way we can query postgres for this information? Using our connection instance

Reply:
> Actually is there a way we can query postgres for this information? 

There is, but it's in a table that tends to change a bit across postgres versions, so I'm reluctant to use it. This is only used for logging so it doesn't have to be perfect. 

> Hmm - this BookmarkLock is shared across all projector threads, right? I'm wondering if we need to worry about thread safety / reliability of this field, even if it's only used for logging. WDYT?

Nice catch. The `synchronized` block is enough, however there's no `synchronized` block in `close()` so that's not safe. Added now.

Reply:
Ahh my bad, I actually forget that we're protected across threads by the synchronised block - ok awesome